velox/common/memory/tests/SharedArbitratorTest.cpp

```diff
@@ -380,6 +380,60 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, queryArbitrationStateCheck) {
   ASSERT_FALSE(queryCtx->testingUnderArbitration());
 }
 
+DEBUG_ONLY_TEST_F(SharedArbitrationTest, raceBetweenAbortAndArbitrationLeave) {
+  const std::vector<RowVectorPtr> vectors =
+      createVectors(rowType_, 32, 32 << 20);
+  setupMemory(kMemoryCapacity, /*memoryPoolInitCapacity=*/0);
+  std::shared_ptr<core::QueryCtx> queryCtx =
+      newQueryCtx(memoryManager_.get(), executor_.get(), 32 << 20);
+
+  folly::EventCount abortWait;
+  std::atomic_bool abortWaitFlag{true};
+  std::atomic<Task*> task{nullptr};
+  const std::string errorMsg{"injected abort error"};
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::Task::leaveSuspended",
+      std::function<void(exec::Task*)>(([&](exec::Task* _task) {
+        if (task.exchange(_task) != nullptr) {
+          return;
+        }
+        abortWaitFlag = false;
+        abortWait.notifyAll();
+        // Let memory pool abort thread to run first. We inject a randomized
+        // delay here to trigger all the potential timing race conditions but
+        // the test result should be the same.
+        std::this_thread::sleep_for(
```

Comment:
Curiously why not use a folly event to ensure the memory pool abort thread runs first?

Reply:
I want to inject some randomness here but expect the same test result. There could be different race combinations here so a random delay can cover them all.